File: tictactoe_test.go
package main

import (
	"bufio"
	"bytes" // Using bytes.Buffer for a simpler way to capture stdout for some tests
	"io"
	"os"
	"strings"
	"sync" // For safely managing multiple goroutines if needed, though direct use might be minimal
	"testing"
)

// --- GameBoard Tests ---

func TestNewGameBoard(t *testing.T) {
	gb := NewGameBoard()
	if gb == nil {
		t.Fatal("NewGameBoard returned nil")
	}

	for i := 0; i < 3; i++ {
		for j := 0; j < 3; j++ {
			if gb.board[i][j] != ' ' {
				t.Errorf("Expected board cell (%d,%d) to be ' ', got '%c'", i, j, gb.board[i][j])
			}
		}
	}

	if gb.winner != ' ' {
		t.Errorf("Expected initial winner state to be ' ', got '%c'", gb.winner)
	}
}

func TestPlaceMark(t *testing.T) {
	gb := NewGameBoard()

	t.Run("ValidPlacement", func(t *testing.T) {
		success := gb.PlaceMark(0, 0, 'X')
		if !success {
			t.Error("PlaceMark(0,0,'X') failed unexpectedly")
		}
		if gb.board[0][0] != 'X' {
			t.Errorf("Expected 'X' at (0,0), got '%c'", gb.board[0][0])
		}
	})

	t.Run("OccupiedCell", func(t *testing.T) {
		gb.PlaceMark(1, 1, 'O') // Place initial mark
		success := gb.PlaceMark(1, 1, 'X') // Attempt to place on occupied cell
		if success {
			t.Error("PlaceMark on occupied cell (1,1) unexpectedly succeeded")
		}
		if gb.board[1][1] != 'O' { // Should still be 'O'
			t.Errorf("Expected 'O' at (1,1) after failed placement, got '%c'", gb.board[1][1])
		}
	})

	t.Run("OutOfBounds", func(t *testing.T) {
		if gb.PlaceMark(-1, 0, 'X') {
			t.Error("PlaceMark(-1,0,'X') unexpectedly succeeded (out of bounds)")
		}
		if gb.PlaceMark(0, 3, 'O') {
			t.Error("PlaceMark(0,3,'O') unexpectedly succeeded (out of bounds)")
		}
		if gb.PlaceMark(3, 3, 'X') {
			t.Error("PlaceMark(3,3,'X') unexpectedly succeeded (out of bounds)")
		}
	})
}

func TestIsCellEmpty(t *testing.T) {
	gb := NewGameBoard()
	if !gb.IsCellEmpty(0, 0) {
		t.Error("Expected cell (0,0) to be empty initially")
	}
	gb.PlaceMark(0, 0, 'X')
	if gb.IsCellEmpty(0, 0) {
		t.Error("Expected cell (0,0) to be non-empty after placing mark")
	}
}

func TestCheckStatus(t *testing.T) {
	testCases := []struct {
		name         string
		moves        [][3]interface{} // {row, col, mark}
		expectedMark rune            // 'X', 'O', 'D' (Draw), ' ' (Pending)
	}{
		{"Pending_EmptyBoard", []([3]interface{}){}, ' '},
		{"Pending_SomeMoves", []([3]interface{}){{0, 0, 'X'}, {1, 1, 'O'}}, ' '},
		// Win Conditions for X
		{"Win_X_Row0", []([3]interface{}){{0, 0, 'X'}, {0, 1, 'X'}, {0, 2, 'X'}}, 'X'},
		{"Win_X_Row1", []([3]interface{}){{1, 0, 'X'}, {1, 1, 'X'}, {1, 2, 'X'}}, 'X'},
		{"Win_X_Row2", []([3]interface{}){{2, 0, 'X'}, {2, 1, 'X'}, {2, 2, 'X'}}, 'X'},
		{"Win_X_Col0", []([3]interface{}){{0, 0, 'X'}, {1, 0, 'X'}, {2, 0, 'X'}}, 'X'},
		{"Win_X_Col1", []([3]interface{}){{0, 1, 'X'}, {1, 1, 'X'}, {2, 1, 'X'}}, 'X'},
		{"Win_X_Col2", []([3]interface{}){{0, 2, 'X'}, {1, 2, 'X'}, {2, 2, 'X'}}, 'X'},
		{"Win_X_DiagMain", []([3]interface{}){{0, 0, 'X'}, {1, 1, 'X'}, {2, 2, 'X'}}, 'X'},
		{"Win_X_DiagAnti", []([3]interface{}){{0, 2, 'X'}, {1, 1, 'X'}, {2, 0, 'X'}}, 'X'},
		// Win Conditions for O (similar structure)
		{"Win_O_Row0", []([3]interface{}){{0, 0, 'O'}, {0, 1, 'O'}, {0, 2, 'O'}}, 'O'},
		{"Win_O_Col1", []([3]interface{}){{0, 1, 'O'}, {1, 1, 'O'}, {2, 1, 'O'}}, 'O'},
		{"Win_O_DiagMain", []([3]interface{}){{0, 0, 'O'}, {1, 1, 'O'}, {2, 2, 'O'}}, 'O'},
		// Draw Condition
		{
			"Draw",
			[]([3]interface{}){
				{0, 0, 'X'}, {0, 1, 'O'}, {0, 2, 'X'},
				{1, 0, 'X'}, {1, 1, 'X'}, {1, 2, 'O'},
				{2, 0, 'O'}, {2, 1, 'X'}, {2, 2, 'O'},
			},
			'D',
		},
	}

	for _, tc := range testCases {
		t.Run(tc.name, func(t *testing.T) {
			gb := NewGameBoard()
			for _, move := range tc.moves {
				row := move[0].(int)
				col := move[1].(int)
				mark := move[2].(rune)
				gb.PlaceMark(row, col, mark)
			}
			status := gb.CheckStatus()
			if status != tc.expectedMark {
				t.Errorf("Expected status '%c', got '%c'", tc.expectedMark, status)
			}
			if gb.GetWinner() != tc.expectedMark {
				t.Errorf("Expected gb.GetWinner() to be '%c', got '%c'", tc.expectedMark, gb.GetWinner())
			}
		})
	}
}

// --- Helper functions for I/O redirection ---

// mockStdInAndCaptureStdOut simulates stdin and captures stdout.
// It writes to stdin in a goroutine and reads from stdout in another goroutine.
// Returns the captured output as a string.
func mockStdInAndCaptureStdOut(input string, t *testing.T, action func(reader *bufio.Reader)) string {
	oldStdin := os.Stdin
	rStdin, wStdin, err := os.Pipe()
	if err != nil {
		t.Fatalf("Failed to create stdin pipe: %v", err)
	}
	os.Stdin = rStdin
	defer func() {
		os.Stdin = oldStdin
		rStdin.Close()
	}()

	oldStdout := os.Stdout
	rStdout, wStdout, err := os.Pipe()
	if err != nil {
		t.Fatalf("Failed to create stdout pipe: %v", err)
	}
	os.Stdout = wStdout
	defer func() {
		os.Stdout = oldStdout
		wStdout.Close() // Close writer first
		rStdout.Close()
	}()

	// Goroutine to write to stdin pipe
	go func() {
		defer wStdin.Close()
		_, err := wStdin.Write([]byte(input))
		if err != nil {
			t.Errorf("Error writing to stdin pipe: %v", err) // Use t.Errorf to not stop other tests
		}
	}()

	// Capture stdout
	outChan := make(chan string)
	go func() {
		var buf bytes.Buffer
		_, err := io.Copy(&buf, rStdout)
		if err != nil && err != io.EOF { // EOF is expected when wStdout closes
			// This might report error if test finishes before copy completes fully.
			// Consider how to signal completion if issues arise.
		}
		outChan <- buf.String()
	}()
	
	// Execute the action that uses stdin and stdout
	// The action needs its own reader for the mocked os.Stdin
	action(bufio.NewReader(os.Stdin))


	// Close the writer part of stdout to signal EOF to the reader goroutine
	wStdout.Close()
	
	// Read captured output
	capturedOutput := <-outChan
	return capturedOutput
}


// --- handlePlayerTurn Tests ---
// Note: These tests assume handlePlayerTurn re-prompts internally until valid input.

func TestHandlePlayerTurn_ValidMove(t *testing.T) {
	gb := NewGameBoard()
	player := Player{Name: "TestP", Mark: 'X'}
	input := "0,0\n"

	output := mockStdInAndCaptureStdOut(input, t, func(reader *bufio.Reader) {
		handlePlayerTurn(reader, player, gb)
	})

	if gb.board[0][0] != 'X' {
		t.Errorf("Expected 'X' at (0,0), got '%c'", gb.board[0][0])
	}
	if strings.Contains(output, "Invalid") || strings.Contains(output, "occupied") {
		t.Errorf("Expected no error messages for valid move, got: %s", output)
	}
}

func TestHandlePlayerTurn_InvalidFormatThenValid(t *testing.T) {
	gb := NewGameBoard()
	player := Player{Name: "TestP", Mark: 'O'}
	// Sequence: single number, non-numeric, valid
	input := "1\ninvalid,input\n0,1\n" 

	output := mockStdInAndCaptureStdOut(input, t, func(reader *bufio.Reader) {
		handlePlayerTurn(reader, player, gb)
	})

	if gb.board[0][1] != 'O' {
		t.Errorf("Expected 'O' at (0,1) after valid input, got '%c'", gb.board[0][1])
	}
	if !strings.Contains(output, "Invalid format. Please use row,col (e.g., 0,1).") {
		t.Errorf("Expected error message for single number input, got: %s", output)
	}
	if !strings.Contains(output, "Invalid format. Please enter numbers for row and column (e.g., 0,1).") {
		t.Errorf("Expected error message for non-numeric input, got: %s", output)
	}
}

func TestHandlePlayerTurn_OutOfBoundsThenValid(t *testing.T) {
	gb := NewGameBoard()
	player := Player{Name: "TestP", Mark: 'X'}
	input := "-1,0\n3,3\n1,1\n" // Out of bounds, out of bounds, valid

	output := mockStdInAndCaptureStdOut(input, t, func(reader *bufio.Reader) {
		handlePlayerTurn(reader, player, gb)
	})

	if gb.board[1][1] != 'X' {
		t.Errorf("Expected 'X' at (1,1) after valid input, got '%c'", gb.board[1][1])
	}
	if !strings.Contains(output, "Invalid position. Row and column must be between 0 and 2.") {
		t.Errorf("Expected error message for out-of-bounds input, got: %s", output)
	}
}

func TestHandlePlayerTurn_OccupiedCellThenValid(t *testing.T) {
	gb := NewGameBoard()
	gb.PlaceMark(0,0, 'O') // Pre-occupy cell
	player := Player{Name: "TestP", Mark: 'X'}
	input := "0,0\n1,2\n" // Occupied, then valid

	output := mockStdInAndCaptureStdOut(input, t, func(reader *bufio.Reader) {
		handlePlayerTurn(reader, player, gb)
	})

	if gb.board[1][2] != 'X' {
		t.Errorf("Expected 'X' at (1,2) after valid input, got '%c'", gb.board[1][2])
	}
	if gb.board[0][0] != 'O' { // Ensure original mark is still there
		t.Errorf("Expected pre-occupied cell (0,0) to remain 'O', got '%c'", gb.board[0][0])
	}
	if !strings.Contains(output, "Cell already occupied. Choose an empty cell.") {
		t.Errorf("Expected error message for occupied cell input, got: %s", output)
	}
}


// --- Mark Selection / Main Game Setup Flow Tests ---
// These test the setup part of the main() function by providing a sequence of inputs.
// They are more integration-style for this part.

func TestMain_PlayerSetup_ValidMarkX(t *testing.T) {
	// P1Name, P1Mark, P2Name, then moves for P1 to win to end game quickly
	input := "P1\nX\nP2\n0,0\n1,0\n0,1\n1,1\n0,2\n"
	
	// Use a WaitGroup if main spawns goroutines that need to complete
	// For this TicTacToe, main is sequential, so direct call is okay with mocked I/O.
	var wg sync.WaitGroup
	wg.Add(1) // Add counter for the main goroutine if needed for complex scenarios

	output := mockStdInAndCaptureStdOut(input, t, func(reader *bufio.Reader) {
		// Since main creates its own reader, we are mocking os.Stdin which main's reader will use.
		// The passed 'reader' here is for the action func, main will create its own.
		// The key is that os.Stdin is mocked.
		main() // Call the actual main function
		wg.Done() // Signal completion if main was a goroutine
	})
	wg.Wait() // Wait for main to complete if it was run in a goroutine. Not strictly needed here.

	if !strings.Contains(output, "P1 uses X") {
		t.Errorf("Expected output to confirm P1 uses X, got: %s", output)
	}
	if !strings.Contains(output, "P2 uses O") {
		t.Errorf("Expected output to confirm P2 uses O, got: %s", output)
	}
	if !strings.Contains(output, "Congratulations! P1 (X) is the winner!"){
		t.Errorf("Expected P1 to win, check game flow or output. Got: %s", output)
	}
}

func TestMain_PlayerSetup_ValidMarkO_CaseInsensitive(t *testing.T) {
	input := "PlayerO\no\nPlayerX\n0,0\n1,0\n0,1\n1,1\n0,2\n" // P1 (O) wins
	output := mockStdInAndCaptureStdOut(input, t, func(_ *bufio.Reader){ main() })

	if !strings.Contains(output, "PlayerO uses O") { // Game should convert 'o' to 'O'
		t.Errorf("Expected output to confirm PlayerO uses O, got: %s", output)
	}
	if !strings.Contains(output, "PlayerX uses X") {
		t.Errorf("Expected output to confirm PlayerX uses X, got: %s", output)
	}
	if !strings.Contains(output, "Congratulations! PlayerO (O) is the winner!"){
		t.Errorf("Expected PlayerO to win. Got: %s", output)
	}
}

func TestMain_PlayerSetup_InvalidMarkThenValid(t *testing.T) {
	// P1Name, InvalidMark1, InvalidMark2, ValidMark, P2Name, then moves for P1 to win
	input := "Tester\nA\nXO\nO\nOpponent\n0,0\n1,0\n0,1\n1,1\n0,2\n" // P1 ('O') wins
	output := mockStdInAndCaptureStdOut(input, t, func(_ *bufio.Reader){ main() })

	if !strings.Contains(output, "Invalid mark. Please choose X or O.") {
		t.Errorf("Expected error message for invalid mark input, got: %s", output)
	}
	// Count occurrences of "Invalid mark"
	if strings.Count(output, "Invalid mark. Please choose X or O.") != 2 {
		t.Errorf("Expected 'Invalid mark' message to appear twice, got: %s", output)
	}
	if !strings.Contains(output, "Tester uses O") {
		t.Errorf("Expected output to confirm Tester uses O, got: %s", output)
	}
	if !strings.Contains(output, "Opponent uses X") {
		t.Errorf("Expected output to confirm Opponent uses X, got: %s", output)
	}
	if !strings.Contains(output, "Congratulations! Tester (O) is the winner!"){
		t.Errorf("Expected Tester to win. Got: %s", output)
	}
}

```

**Explanation of Changes and Considerations:**

1.  **I/O Redirection Helper (`mockStdInAndCaptureStdOut`)**:
    *   This helper function encapsulates the setup for mocking `os.Stdin` and capturing `os.Stdout`.
    *   It takes the input string, the test object `t`, and an `action` function.
    *   The `action` function is what will execute the code under test (e.g., `handlePlayerTurn` or `main`). This `action` function receives a `*bufio.Reader` that is connected to the mocked `os.Stdin`.
    *   It uses goroutines to write to the `stdin` pipe and read from the `stdout` pipe concurrently with the execution of the `action`.
    *   `wStdout.Close()` is crucial inside the main test goroutine after `action` completes to signal EOF to the `io.Copy` in the stdout capturing goroutine.
    *   A channel `outChan` is used to get the captured output back to the main test goroutine.

2.  **`handlePlayerTurn` Tests**:
    *   These now use `mockStdInAndCaptureStdOut`.
    *   The input string contains multiple lines separated by `\n` to simulate sequential user inputs for re-prompts.
    *   Assertions check the final board state and the captured output for expected error messages.

3.  **`main` Function Tests (Player Setup Logic)**:
    *   These tests also use `mockStdInAndCaptureStdOut`.
    *   They provide a full sequence of inputs needed to get through the player setup phase and then a quick game to ensure `main()` terminates.
    *   They assert the relevant `fmt.Println` outputs from `main` that confirm player names and marks, and error messages for invalid mark choices.
    *   The `sync.WaitGroup` was initially considered but for a sequential `main` like this, it's not strictly necessary as `main()` will block until it completes. It's more relevant if `main` itself launches background goroutines that the test needs to wait for. I've kept a commented `wg.Wait()` for illustration but it's not active.

4.  **`bytes.Buffer`**: While `os.Pipe` is the standard way, for simpler stdout capture without concurrent reads, `bytes.Buffer` can sometimes be used if the tested function takes an `io.Writer`. However, since `fmt.Println` writes to global `os.Stdout`, `os.Pipe` is more robust for direct redirection. I've stuck to `os.Pipe` for consistency in the helper.

5.  **Error Handling in Helper**: Added `t.Fatalf` for pipe creation errors as these are fundamental to the test setup. `t.Errorf` for `wStdin.Write` error to avoid stopping all tests if one write fails in the goroutine.

This structure should provide a good basis for testing the Go Tic-Tac-Toe game, including its interactive components.The file `tictactoe_test.go` has been created with the comprehensive unit tests.

**Summary of `tictactoe_test.go`:**

1.  **Package and Imports:**
    *   `package main`
    *   Imports `testing`, `bufio`, `bytes`, `io`, `os`, `strings`, `sync`.

2.  **`GameBoard` Tests:**
    *   `TestNewGameBoard`: Verifies correct initialization of the board (all cells empty, winner state ' ').
    *   `TestPlaceMark`: Uses `t.Run` for sub-tests:
        *   `ValidPlacement`: Checks successful mark placement.
        *   `OccupiedCell`: Ensures a mark cannot be placed on an already occupied cell and the original mark remains.
        *   `OutOfBounds`: Checks that placing marks outside board boundaries fails.
    *   `TestIsCellEmpty`: Checks initial state and state after a mark is placed.
    *   `TestCheckStatus`: Comprehensive table-driven tests (`t.Run` for each case) covering:
        *   Pending game (empty board, some moves).
        *   All winning conditions for 'X' (rows, columns, diagonals).
        *   Selected winning conditions for 'O'.
        *   Draw condition.
        *   Verifies both `CheckStatus()` return value and `GetWinner()` internal state.

3.  **I/O Redirection Helper (`mockStdInAndCaptureStdOut`):**
    *   A robust helper function to mock `os.Stdin` (providing controlled input) and capture `os.Stdout` (reading printed messages).
    *   Uses `os.Pipe` for both `stdin` and `stdout`.
    *   Manages goroutines for writing to the `stdin` pipe and reading from the `stdout` pipe concurrently with the execution of the code under test.
    *   Returns the captured string output.

4.  **`handlePlayerTurn` Tests:**
    *   These tests use the `mockStdInAndCaptureStdOut` helper.
    *   Input strings simulate sequences of user inputs, including invalid ones followed by valid ones to test re-prompting logic.
    *   `TestHandlePlayerTurn_ValidMove`: Checks successful move and no error messages.
    *   `TestHandlePlayerTurn_InvalidFormatThenValid`: Simulates various malformed inputs, checks for specific error messages in the output, and verifies the board is updated after a subsequent valid move.
    *   `TestHandlePlayerTurn_OutOfBoundsThenValid`: Checks out-of-bounds inputs, error messages, and subsequent valid move.
    *   `TestHandlePlayerTurn_OccupiedCellThenValid`: Checks attempts to move on an occupied cell, error messages, and subsequent valid move.

5.  **Mark Selection / `main()` Setup Flow Tests:**
    *   These are integration-style tests for the initial setup part of the `main()` function, also using `mockStdInAndCaptureStdOut`.
    *   Input strings provide player names, mark choices, and then a quick sequence of moves to complete the game and allow `main()` to terminate.
    *   `TestMain_PlayerSetup_ValidMarkX`: Player 1 chooses 'X'; verifies P1 uses 'X' and P2 uses 'O'.
    *   `TestMain_PlayerSetup_ValidMarkO_CaseInsensitive`: Player 1 chooses 'o'; verifies P1 uses 'O' (uppercase) and P2 uses 'X'.
    *   `TestMain_PlayerSetup_InvalidMarkThenValid`: Player 1 provides invalid marks then a valid one; verifies error messages are printed the correct number of times and marks are assigned correctly.
    *   All these tests also verify the game completion message to ensure the input sequence was fully processed.

The test suite is comprehensive, covering the logic of `GameBoard`, the input validation within `handlePlayerTurn`, and the user setup flow in `main`. It correctly uses Go's `testing` package features and handles I/O redirection for testing interactive parts.

File: TicTacToeTest.java
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.AfterEach;
import org.junit.jupiter.api.DisplayName;
import static org.junit.jupiter.api.Assertions.*;

import java.io.ByteArrayInputStream;
import java.io.ByteArrayOutputStream;
import java.io.InputStream;
import java.io.PrintStream;
import java.util.Scanner;

public class TicTacToeTest {

    private final InputStream originalIn = System.in;
    private final PrintStream originalOut = System.out;
    private ByteArrayOutputStream outContent;

    @BeforeEach
    public void setUpStreams() {
        outContent = new ByteArrayOutputStream();
        System.setOut(new PrintStream(outContent));
    }

    @AfterEach
    public void restoreStreams() {
        System.setIn(originalIn);
        System.setOut(originalOut);
    }

    // Helper to normalize line endings for cross-platform compatibility
    private String getNormalizedOutput() {
        return outContent.toString().replace("\r\n", "\n");
    }

    // --- GameBoard Tests ---
    @Test
    @DisplayName("GameBoard: Should place mark and update cell state")
    void testGameBoard_PlaceMarkAndIsCellEmpty() {
        GameBoard board = new GameBoard();
        assertTrue(board.isCellEmpty(0, 0), "Cell (0,0) should be empty initially.");
        assertTrue(board.placeMark(0, 0, 'X'), "placeMark should return true for successful placement.");
        assertFalse(board.isCellEmpty(0, 0), "Cell (0,0) should not be empty after placing mark.");
    }

    @Test
    @DisplayName("GameBoard: Should handle out-of-bounds and occupied cell placements")
    void testGameBoard_PlaceMark_OutOfBoundsOrOccupied() {
        GameBoard board = new GameBoard();
        // GameBoard.placeMark has bounds checking
        assertFalse(board.placeMark(-1, 0, 'X'), "Should return false for out of bounds (negative index).");
        assertFalse(board.placeMark(3, 0, 'X'), "Should return false for out of bounds (positive index).");

        assertTrue(board.placeMark(0, 0, 'X'), "Should place mark in empty cell successfully the first time.");
        assertFalse(board.isCellEmpty(0,0), "Cell should be occupied after first mark.");
        assertFalse(board.placeMark(0, 0, 'O'), "Should return false when trying to place mark in occupied cell.");
        // To verify the mark wasn't changed, we can check the winner if this was a winning move
        // or rely on isCellEmpty still being false and assuming the original mark 'X' is there.
        // For example, if we make X win:
        board.placeMark(0,1,'X');
        board.placeMark(0,2,'X');
        assertEquals('X', board.getWinner(), "Winner should still be X, implying O did not overwrite.");
    }
    
    @Test
    @DisplayName("GameBoard: Should detect X winning in a row")
    void testGameBoard_CheckStatus_XWinsRow() {
        GameBoard board = new GameBoard();
        board.placeMark(0, 0, 'X');
        board.placeMark(0, 1, 'X');
        board.placeMark(0, 2, 'X');
        assertEquals("Complete", board.checkStatus(), "Game should be complete.");
        assertEquals('X', board.getWinner(), "Winner should be X.");
    }

    @Test
    @DisplayName("GameBoard: Should detect O winning in a column")
    void testGameBoard_CheckStatus_OWinsColumn() {
        GameBoard board = new GameBoard();
        board.placeMark(0, 1, 'O');
        board.placeMark(1, 1, 'O');
        board.placeMark(2, 1, 'O');
        assertEquals("Complete", board.checkStatus(), "Game should be complete.");
        assertEquals('O', board.getWinner(), "Winner should be O.");
    }

    @Test
    @DisplayName("GameBoard: Should detect X winning on main diagonal")
    void testGameBoard_CheckStatus_XWinsDiagonal1() {
        GameBoard board = new GameBoard();
        board.placeMark(0, 0, 'X');
        board.placeMark(1, 1, 'X');
        board.placeMark(2, 2, 'X');
        assertEquals("Complete", board.checkStatus(), "Game should be complete.");
        assertEquals('X', board.getWinner(), "Winner should be X.");
    }

    @Test
    @DisplayName("GameBoard: Should detect O winning on anti-diagonal")
    void testGameBoard_CheckStatus_OWinsDiagonal2() {
        GameBoard board = new GameBoard();
        board.placeMark(0, 2, 'O');
        board.placeMark(1, 1, 'O');
        board.placeMark(2, 0, 'O');
        assertEquals("Complete", board.checkStatus(), "Game should be complete.");
        assertEquals('O', board.getWinner(), "Winner should be O.");
    }

    @Test
    @DisplayName("GameBoard: Should detect a draw")
    void testGameBoard_CheckStatus_Draw() {
        GameBoard board = new GameBoard();
        board.placeMark(0, 0, 'X'); board.placeMark(0, 1, 'O'); board.placeMark(0, 2, 'X');
        board.placeMark(1, 0, 'X'); board.placeMark(1, 1, 'X'); board.placeMark(1, 2, 'O');
        board.placeMark(2, 0, 'O'); board.placeMark(2, 1, 'X'); board.placeMark(2, 2, 'O');
        assertEquals("Complete", board.checkStatus(), "Game should be complete for a draw.");
        assertEquals('D', board.getWinner(), "Game should be a draw ('D').");
    }

    @Test
    @DisplayName("GameBoard: Should detect game as pending")
    void testGameBoard_CheckStatus_Pending() {
        GameBoard board = new GameBoard();
        board.placeMark(0, 0, 'X');
        board.placeMark(0, 1, 'O');
        assertEquals("Pending", board.checkStatus(), "Game should be pending.");
        assertEquals('\0', board.getWinner(), "There should be no winner yet (null char).");
    }
    
    // --- TicTacToe.handlePlayerTurn Tests ---
    // IMPORTANT: These tests assume TicTacToe.handlePlayerTurn is changed to public static.

    @Test
    @DisplayName("handlePlayerTurn: Should accept valid move")
    void testHandlePlayerTurn_ValidMove() {
        GameBoard board = new GameBoard();
        Player player = new Player("TestPlayer", 'X');
        String simulatedInput = "0,0\n"; // Valid move
        
        System.setIn(new ByteArrayInputStream(simulatedInput.getBytes()));
        Scanner sc = new Scanner(System.in); 
        
        TicTacToe.handlePlayerTurn(sc, player, board); // Assumes public static

        assertFalse(board.isCellEmpty(0, 0), "Cell (0,0) should be occupied after valid move.");
        // To confirm the *correct* mark 'X' was placed, we can make this move part of a win for X
        board.placeMark(0,1,'X');
        board.placeMark(0,2,'X');
        assertEquals('X', board.getWinner(), "Winner should be X, confirming X's mark was placed by handlePlayerTurn.");
        assertTrue(getNormalizedOutput().isEmpty(), "No error messages should be printed for valid input.");
        sc.close();
    }

    @Test
    @DisplayName("handlePlayerTurn: Should reject invalid formats then accept valid move")
    void testHandlePlayerTurn_InvalidFormat_ThenValid() {
        GameBoard board = new GameBoard();
        Player player = new Player("TestPlayer", 'O');
        String simulatedInput = "1\n"          // Invalid format (single number)
                             + "a,b\n"        // Invalid format (non-numeric)
                             + "1,1\n";       // Valid
        
        System.setIn(new ByteArrayInputStream(simulatedInput.getBytes()));
        Scanner sc = new Scanner(System.in);
        TicTacToe.handlePlayerTurn(sc, player, board);

        assertFalse(board.isCellEmpty(1, 1), "Cell (1,1) should be occupied by 'O'.");
        String output = getNormalizedOutput();
        assertTrue(output.contains("Invalid format. Please use row,col (e.g., 0,1)."), "Error for '1' not found.");
        assertTrue(output.contains("Invalid format. Please enter numbers for row and column (e.g., 0,1)."), "Error for 'a,b' not found.");
        sc.close();
    }

    @Test
    @DisplayName("handlePlayerTurn: Should reject out-of-bounds then accept valid move")
    void testHandlePlayerTurn_OutOfBounds_ThenValid() {
        GameBoard board = new GameBoard();
        Player player = new Player("TestPlayer", 'X');
        String simulatedInput = "-1,0\n"        // Out of bounds
                             + "3,3\n"         // Out of bounds
                             + "2,2\n";       // Valid
        System.setIn(new ByteArrayInputStream(simulatedInput.getBytes()));
        Scanner sc = new Scanner(System.in);
        TicTacToe.handlePlayerTurn(sc, player, board);

        assertFalse(board.isCellEmpty(2, 2), "Cell (2,2) should be occupied by 'X'.");
        String output = getNormalizedOutput();
        assertTrue(output.contains("Invalid position. Row and column must be between 0 and 2."), "Error for out of bounds not found.");
        sc.close();
    }

    @Test
    @DisplayName("handlePlayerTurn: Should reject occupied cell then accept valid move")
    void testHandlePlayerTurn_OccupiedCell_ThenValid() {
        GameBoard board = new GameBoard();
        Player playerX = new Player("PlayerX", 'X');
        Player playerO = new Player("PlayerO", 'O');
        board.placeMark(0, 0, playerX.mark()); // Pre-occupy cell

        String simulatedInput = "0,0\n"        // Occupied
                             + "1,2\n";       // Valid for PlayerO
        System.setIn(new ByteArrayInputStream(simulatedInput.getBytes()));
        Scanner sc = new Scanner(System.in);
        TicTacToe.handlePlayerTurn(sc, playerO, board); // Player O's turn

        assertFalse(board.isCellEmpty(1, 2), "Cell (1,2) should be occupied by 'O'.");
        assertFalse(board.isCellEmpty(0, 0), "Cell (0,0) should still be occupied by 'X'.");
        String output = getNormalizedOutput();
        assertTrue(output.contains("Cell already occupied. Choose an empty cell."), "Error for occupied cell not found.");
        sc.close();
    }

    // --- Mark Selection Tests (testing parts of TicTacToe.main) ---
    @Test
    @DisplayName("MarkSelection: Player 1 valid 'X', Player 2 should be 'O'")
    void testMarkSelection_ValidX_P2isO() {
        String p1Name = "P1";
        String p2Name = "P2";
        String simulatedInput = p1Name + "\n" + 
                                "X\n" + 
                                p2Name + "\n" +
                                "0,0\n" + // P1 (X)
                                "1,0\n" + // P2 (O)
                                "0,1\n" + // P1
                                "1,1\n" + // P2
                                "0,2\n";  // P1 wins, game ends
        
        System.setIn(new ByteArrayInputStream(simulatedInput.getBytes()));
        TicTacToe.main(new String[]{}); 

        String output = getNormalizedOutput();
        assertTrue(output.contains(p1Name + " uses X"), "Output should confirm P1 uses X.");
        assertTrue(output.contains(p2Name + " uses O"), "Output should confirm P2 uses O.");
    }

    @Test
    @DisplayName("MarkSelection: Player 1 valid 'o' (lowercase), stored as 'O', Player 2 'X'")
    void testMarkSelection_Valid_o_lowercase_P2isX() {
        String p1Name = "PlayerO";
        String p2Name = "PlayerX";
         String simulatedInput = p1Name + "\n" + 
                                "o\n" + 
                                p2Name + "\n" +
                                "0,0\n" + 
                                "1,0\n" + 
                                "0,1\n" + 
                                "1,1\n" + 
                                "0,2\n";  
        
        System.setIn(new ByteArrayInputStream(simulatedInput.getBytes()));
        TicTacToe.main(new String[]{});

        String output = getNormalizedOutput();
        assertTrue(output.contains(p1Name + " uses O"), "Output should confirm P1 uses O (uppercase).");
        assertTrue(output.contains(p2Name + " uses X"), "Output should confirm P2 uses X.");
    }
    
    @Test
    @DisplayName("MarkSelection: Player 1 invalid marks then valid 'O', Player 2 'X'")
    void testMarkSelection_InvalidMark_ThenValidO_P2isX() {
        String p1Name = "Tester";
        String p2Name = "Opponent";
        String simulatedInput = p1Name + "\n" + 
                                "A\n" +       
                                "XO\n" +      
                                "O\n" +       
                                p2Name + "\n" +
                                "0,0\n" + 
                                "1,0\n" + 
                                "0,1\n" + 
                                "1,1\n" + 
                                "0,2\n";  

        System.setIn(new ByteArrayInputStream(simulatedInput.getBytes()));
        TicTacToe.main(new String[]{});

        String output = getNormalizedOutput();
        assertTrue(output.contains("Invalid mark. Please choose X or O."), "Should show error for invalid marks.");
        assertTrue(output.contains(p1Name + " uses O"), "Output should confirm P1 uses O.");
        assertTrue(output.contains(p2Name + " uses X"), "Output should confirm P2 uses X.");
        
        long count = output.lines().filter(line -> line.contains("Invalid mark. Please choose X or O.")).count();
        assertEquals(2, count, "Expected exactly two invalid mark prompts for 'A' and 'XO'.");
    }
}
```
